modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotTraverse.scala

```diff
@@ -0,0 +1,82 @@
+package org.tessellation.infrastructure.snapshot
+
+import cats._
+import cats.syntax.applicative._
+import cats.syntax.either._
+import cats.syntax.flatMap._
+import cats.syntax.functor._
+
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema._
+import org.tessellation.sdk.domain.snapshot.SnapshotContextFunctions
+import org.tessellation.security.hash.Hash
+import org.tessellation.security.signature.Signed
+
+import higherkindness.droste._
+import higherkindness.droste.data._
+import higherkindness.droste.util.DefaultTraverse
+
+sealed trait StackF[A]
+
+case class More[A](a: A, step: Signed[GlobalIncrementalSnapshot]) extends StackF[A]
+case class Done[A](result: GlobalSnapshot) extends StackF[A]
+
+object StackF {
+
+  implicit val traverse: Traverse[StackF] = new DefaultTraverse[StackF] {
+    override def traverse[G[_]: Applicative, A, B](fa: StackF[A])(f: A => G[B]): G[StackF[B]] =
+      fa match {
+        case More(a, step) => f(a).map(More(_, step))
+        case Done(r)       => (Done(r): StackF[B]).pure[G]
+      }
+  }
+}
+
+trait GlobalSnapshotTraverse[F[_]] {
+  def computeState(latest: Signed[GlobalIncrementalSnapshot]): F[GlobalSnapshotInfo]
+}
+
+object GlobalSnapshotTraverse {
```

Comment:
I think we can consider rewriting it using fs2 but I'd opt for doing that later (but as high prio ticket)

Reply:
I don't think we should do it later since we're creating technical debt out of thin air. It's a simple change and I would be surprised if it would be more than 20 LOC.

Reply:
anyway, I don't think it's overengineered. you specify just two things:
1. logic for loading snapshot
2. logic for calculating state

and that's all. recursion is done automatically by the library.

Reply:
I understand how it works, it doesn't change the fact that it's a lot of code for not much of functionality.

Code below is doing the same thing in less code than `GlobalSnapshotTraverse` and `GlobalSnapshotLoader` (and also partially `Main` - no need to load the latest snapshot in `Main`). Also, this implementation doesn't load all snapshots into memory at once.

```scala
def loadChain[F[_]: Monad: MonadThrow](
  contextFns: SnapshotContextFunctions[F, GlobalSnapshotArtifact, GlobalSnapshotContext],
  loadInc: Hash => F[Option[Signed[GlobalIncrementalSnapshot]]],
  loadFull: Hash => F[Option[Signed[GlobalSnapshot]]],
  rollbackHash: Hash
): F[(GlobalSnapshotInfo, GlobalIncrementalSnapshot)] = {

  def loadFullOrErr: Hash => F[Signed[GlobalSnapshot]] =
    loadFull(_).flatMap(_.liftTo[F](new Throwable(s"Global snapshot not found, hash=${rollbackHash.show}")))

  def loadIncOrErr: Hash => F[Signed[GlobalIncrementalSnapshot]] =
    loadInc(_).flatMap(_.liftTo[F](new Throwable(s"Incremental snapshot not found, hash=${rollbackHash.show}")))

  def hashChain(hash: Hash): F[NonEmptyChain[Hash]] =
    loadInc(hash).flatMap {
      _.traverse { inc =>
        hashChain(inc.lastSnapshotHash).map(_.append(hash))
      }.map(_.getOrElse(NonEmptyChain.one(hash)))
    }

  for {
    rollbackInc <- loadIncOrErr(rollbackHash)
    (globalHash, incHashesNec) <- hashChain(rollbackInc.lastSnapshotHash).map { nec =>
      (nec.head, NonEmptyChain.fromChainAppend(nec.tail, rollbackHash))
    }
    global <- loadFullOrErr(globalHash)
    firstInc <- loadIncOrErr(incHashesNec.head)

    (info, lastInc) <- incHashesNec.tail.foldLeftM((toGlobalSnapshotInfo(global.info), firstInc)) { (acc, hash) =>
      loadIncOrErr(hash).flatMap { inc =>
        contextFns.createContext(acc._1, acc._2, inc).map(_ -> inc)
      }
    }
  } yield (info, lastInc)
}

```

Reply:
done :+1: